### PR TITLE
Dueling picklist

### LINKED ISSUE #1667: Add Dueling Picklist
I've implemented a new [`<DuelingPicklist />`](https://www.lightningdesignsystem.com/components/dueling-picklist/) component for my team's internal project (Salesforce employee), which I would like to contribute. 

Proposed `propTypes`:
```
static propTypes = {
    /**
     * allows the user to reorder the second listbox of options
     */
    allowReordering: PropTypes.bool,

    /**
     * **Assistive text for accessibility**
     * This object is merged with the default props object on every render.
     * * `optionDragLabel`: Instructions on how to drag and drop with a keyboard.
     * * `moveSelectionUp`: Used by "up" reordering button.
     * * `moveSelectionDown`: Used by "down" reordering button.
     * * `moveSelectionToSecondCategory`: Used by "right" button, which moves selected items to selection.
     * * `moveSelectionToFirstCategory`: Used by "left" button, which removes selected items from selection.
     * * `tooltipIcon`: Used by tooltip, if a tooltip is being used.
     * * `itemLocked`: Used to label locked items.
     * * `itemsMovedToSelection`: Used in Aria Live area to inform user that items were moved to selection.
     * * `itemsRemovedFromSelection`: Used in Aria Live area to inform user that items were removed from selection..
     * * `itemsReorderedInSelection`: Used in Aria Live area to inform user that items were reordered in selection.
     */ 
    assistiveText: PropTypes.shape({
        optionDragLabel: PropTypes.string,
        moveSelectionUp: PropTypes.string,
        moveSelectionDown: PropTypes.string,
        moveSelectionToSecondCategory: PropTypes.string,
        moveSelectionToFirstCategory: PropTypes.string,
        tooltipIcon: PropTypes.string,
        itemLocked: PropTypes.string,
        itemsMovedToSelection: PropTypes.string,
        itemsRemovedFromSelection: PropTypes.string,
        itemsReorderedInSelection: PropTypes.string,
    }),

    /**
     * When true, the height of both listboxes will be the smallest height needed to show all items without having to scroll.
     */ 
    automaticHeightMinimization: PropTypes.bool,

    /**
     * When true, all interactions are disabled.
     */
    disabled: PropTypes.bool,
    
    /**
     * Event Callbacks
     * * `onChange`: Called when items are added or removed from `selection`
     */
    events: PropTypes.shape({
        onChange: PropTypes.func.isRequired,
    }).isRequired,

    /**
     * When true, shows a tooltip.
     */
    hasTooltip: PropTypes.bool,

    /**
     * Manually sets the height of both listboxes.
     */
    height: PropTypes.string,

    /**
     * Element ids (used for accessibility). If not provided, ids will be generated with shortid.
     * * `picklistGroupLabel`: id for labeling the `<DuelingPicklist />` component.
     * * `dragLiveRegion`: id for Aria Live element.
     * * `optionDragLabel`: id for describing how to use keyboard interactions.
     * * `firstCategoryLabel`: id for options listbox.
     * * `secondCategoryLabel`: id for selection listbox.
     */
    ids: PropTypes.shape({
        picklistGroupLabel: PropTypes.string,
        dragLiveRegion: PropTypes.string,
        optionDragLabel: PropTypes.string,
        firstCategoryLabel: PropTypes.string,
        secondCategoryLabel: PropTypes.string,
    }),

    /**
     * Labels
     * * `heading`: Heading for component.
     * * `firstCategory`: Label for options.
     * * `secondCategory`: Label for selection.
     * * `tooltip`: Label for tooltip.
     * * `selectedItems`: Labels selected items. Used in View Mode.
     */
    labels: PropTypes.shape({
        heading: PropTypes.string,
        firstCategory: PropTypes.string,
        secondCategory: PropTypes.string,
        tooltip: PropTypes.string,
        selectedItems: PropTypes.string,
    }),

    /**
     * Items in the first listbox
     * * `label`: Item label.
     * * `id`: Unique id for the item.
     * * `locked`: When true, a lock icon renders on the item, and the item cannot be moved to or from the selection.
     */
    options: PropTypes.arrayOf(PropTypes.shape({
        label: PropTypes.string.isRequired,
        id: PropTypes.string.isRequired,
        locked: PropTypes.bool,
    })).isRequired,

    /**
     * When true, the component will be render with responsive css classes applied. Any items longer than the space available will truncate with ellipses.
     */
    responsive: PropTypes.bool,

    /**
     * When true, a red asterisk will render, visually marking the item as required.
     */
    required: PropTypes.bool,

    /**
     * Items in the second listbox
     * * `label`: Item label.
     * * `id`: Unique id for the item.
     * * `locked`: When true, a lock icon renders on the item, and the item cannot be moved to or from the selection.
     */
    selection: PropTypes.arrayOf(PropTypes.shape({
        label: PropTypes.string.isRequired,
        id: PropTypes.string.isRequired,
        locked: PropTypes.bool,
    })).isRequired,

    /**
     * When true, the component will render in View Mode, which only shows the items in the selection.
     */
    viewMode: PropTypes.bool,
}
```

### PR DESCRIPTION
**Pull request by kyleheddon**
Fixes #1667 

### Additional description
Implements [`<DuelingPicklist />`](https://lightningdesignsystem.com/components/dueling-picklist/)

[Storybook examples](https://design-system-react-co-pr-1712.herokuapp.com/?selectedKind=SLDSDuelingPicklist&selectedStory=Default&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook%2Factions%2Factions-panel)

---

### CONTRIBUTOR checklist (do not remove)
Please complete for every pull request

* [x] First-time contributors should sign the Contributor License Agreement. It's a fancy way of saying that you are giving away your contribution to this project. If you haven't before, wait a few minutes and a bot will comment on this pull request with instructions.
* [x] `npm run lint:fix` has been run and linting passes.
* [x] Mocha, Jest (Storyshots), and `components/component-docs.json` CI tests pass (`npm test`).
* [x] Tests have been added for new props to prevent regressions in the future. See [readme](https://github.com/salesforce/design-system-react/blob/master/tests/README.md).
* [x] Review the appropriate Storybook stories. Open [http://localhost:9001/](http://localhost:9001/).
* [x] The Accessibility panel of each Storybook story has 0 violations (aXe). Open [http://localhost:9001/](http://localhost:9001/).
* [x] Review tests are passing in the browser. Open [http://localhost:8001/](http://localhost:8001/).
* [x] Review markup conforms to [SLDS](https://www.lightningdesignsystem.com/) by looking at [DOM snapshot strings](https://facebook.github.io/jest/docs/en/snapshot-testing.html).

### REVIEWER checklist (do not remove)

* [ ] TravisCI tests pass. This includes linting, Mocha, Jest, Storyshots, and `components/component-docs.json` tests.
* [ ] Tests have been added for new props to prevent regressions in the future. See [readme](https://github.com/salesforce/design-system-react/blob/master/tests/README.md).
* [ ] Review the appropriate Storybook stories. Open [http://localhost:9001/](http://localhost:9001/).
* [ ] The Accessibility panel of each Storybook story has 0 violations (aXe). Open [http://localhost:9001/](http://localhost:9001/).
* [ ] Review tests are passing in the browser. Open [http://localhost:8001/](http://localhost:8001/).
* [ ] Review markup conforms to [SLDS](https://www.lightningdesignsystem.com/) by looking at [DOM snapshot strings](https://facebook.github.io/jest/docs/en/snapshot-testing.html).
###### Required only if there are markup / UX changes
* [ ] Add year-first date and commit SHA to `last-slds-markup-review` in `package.json` and push.
* [ ] Request a review of the deployed Heroku app by the Salesforce UX Accessibility Team.
* [ ] Add year-first review date, and commit SHA, `last-accessibility-review`, to `package.json` and push.
* [ ] While the contributor's branch is checked out, run `npm run local-update` within locally cloned [site repo](https://github.com/salesforce-ux/design-system-react-site) to confirm the site will function correctly at the next release.

### CODE REVIEW

**Comment by interactivellama** [comment]
I've set this up to auto-deploy the branch on push in case you want to send a link out in the future.

https://design-system-react-co-pr-1712.herokuapp.com/?selectedKind=SLDSDuelingPicklist&selectedStory=Default&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook%2Factions%2Factions-panel

File: components/dueling-picklist/private/option.jsx [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+};
+
+const optionSourceCollector = (connect, monitor) => ({
+	connectDragSource: connect.dragSource(),
+	isDragging: monitor.isDragging(),
+});
+
+const optionTarget = {
+	drop (props, monitor, component) {
+		const hoverBoundingRect = findDOMNode(component).getBoundingClientRect();
```
Comment (kyleheddon) [inline components/dueling-picklist/private/option.jsx]: I'm getting this eslint warning:
```
error  Do not use findDOMNode                                    react/no-find-dom-node
```

@interactivellama, do you know of an alternative approach for detecting Drag and Drop boundaries?

**Comment by kyleheddon** [comment]
@interactivellama I think I'm ready for a review. I just submitted the CLA form.

**Comment by amateurhuman** [comment]
@kyleheddon I've added you to the Salesforce org, once you accept the invite you can refresh the CLA check by going to https://cla.salesforce.com/status/salesforce/design-system-react/pull/1712

**Comment by kyleheddon** [comment]
I'm not sure how to resolve the build issues. `npm test` and `npm run travis-ci` both pass locally 🤷‍♂️

**Comment by interactivellama** [comment]
No problem. I'll look into to it.

**Comment by kyleheddon** [comment]
Dang, not sure how those 2 files got missed. I've run `npm run lint:fix` quite a few times, but must have made some changes after linting. My apologies. Thanks for checking it out. It could be eslint-config-prettier version issues - not sure.

**Comment by kyleheddon** [comment]
I see what happened. I made a last minute change to a prop name, which caused a line length to go over, and didn't run lint fix afterwards. Confusion was added when my tests passed.
screen shot 2019-01-25 at 12 59 52 pm

**Comment by interactivellama** [comment]
@kyleheddon This is looking amazing so far!

UX side of review: I haven't found any issues. SLDS pattern spec appears to be fully implemented. I'll look at accessibility and the code this afternoon.

**Comment by interactivellama** [comment]
#### Accessibility notes
* I'm hearing the screenreader say "undefined moved to Second Category". The text concatenation for the assertive text appears to need to reference the option in the old location. This might be a good place for internal state of previous selection--that is to store the current selection in state when the move is pressed.

Otherwise this looks good. I'll look at the code next.

**Comment by kyleheddon** [comment]
> #### Accessibility notes
> * I'm hearing the screenreader say "undefined moved to Second Category".

Can you give me the steps to reproduce? So far I haven't been able to recreate

**Comment by interactivellama** [comment]
I'm getting it by moving an option from category one to category two with the > "right" button. I'm guessing the re-render of UI is off from the "state" of the options.

File: components/dueling-picklist/__examples__/automatic-height-minimization.jsx [hunk 2]
```diff
@@ -0,0 +1,56 @@
+/* eslint-disable no-console, react/prop-types */
+import React from 'react';
+import IconSettings from '~/components/icon-settings';
+import DuelingPicklist from '~/components/dueling-picklist';
+import duelingPicklistFilter from '~/components/dueling-picklist/filter';
```
Comment (interactivellama) [inline components/dueling-picklist/__examples__/automatic-height-minimization.jsx]: This needs to be exported in the `components/index.js`. Examples only have access to exports in that file.

File: components/dueling-picklist/private/options.jsx [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+		let tabIndex = '-1';
+		if (!disabled && (isInFocus || (noItemsSelected && index === 0))) {
+			tabIndex = '0';
+		}
+
+		return (
+			<Option
+				option={option}
+				key={index}
```
[the remark this answers is not in the document]
Comment (kyleheddon) [reply components/dueling-picklist/private/options.jsx]: that make a lot of sense

**Comment by kyleheddon** [comment]
> I might go through and pull request some code comments into your branch, if you are open to that.

I'm very open to that. I imagine there could be a lot to learn from this.

**Comment by tanhengyeow** [comment]
@kyleheddon @interactivellama Are you guys working on this? I would be more than happy to move this forward and learn from the current work :)

**Comment by jesseditson** [comment]
Does this need a revisit? At Quip we have an implementation of this as well that we could consider promoting if this is not mergeable/active, otherwise we'd love to converge on the upstream version if possible.

**Comment by interactivellama** [comment]
@jesseditson Sounds good. It was abandoned a while back and not picked up again.

**Pull request closed without merging**